### PR TITLE
[Reviewer: Andy] Allow IPv6 DNS servers

### PR DESCRIPTION
**Pull request by rkd-msw**
This fixes https://github.com/Metaswitch/cpp-common/issues/246, mostly by porting code from Sprout's ENUM code.

I've given the resolver three constructors - one with a single IP address as a string, one with multiple IP address as strings, and one taking IP46Addresses directly (so that the ENUM resolution code can plug into this).

### CODE REVIEW

File: src/dnscachedresolver.cpp [hunk 1]
```diff
@@ -149,34 +135,39 @@ DnsCachedResolver::DnsCachedResolver(const std::vector<std::string>& dns_servers
...
+    }
+    else if (inet_pton(AF_INET6, dns_servers[i].c_str(), &(addr.addr.ipv6)))
+    {
+      addr.af = AF_INET6;
+    }
+    else
+    {
+      TRC_ERROR("Failed to parse '%s' as IP address - defaulting to 127.0.0.1", dns_servers[i].c_str());
+      addr.af = AF_INET;
+      (void)inet_aton("127.0.0.1", &(addr.addr.ipv4));
```
Comment (bossmc) [inline src/dnscachedresolver.cpp]: Does it matter if this appears multiple times in the final list?

File: src/dnscachedresolver.cpp [hunk 2]
```diff
@@ -149,34 +135,39 @@ DnsCachedResolver::DnsCachedResolver(const std::vector<std::string>& dns_servers
   pthread_condattr_destroy(&cond_attr);
 }
 
-DnsCachedResolver::DnsCachedResolver(const std::string& dns_server) :
-  _cache_lock(PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP),
-  _cache()
+DnsCachedResolver* DnsCachedResolver::from_server_ips(const std::vector<std::string>& dns_servers)
 {
-  TRC_STATUS("Creating Cached Resolver using server %s", dns_server.c_str());
+  std::vector<IP46Address> dns_server_ips;
```
Comment (bossmc) [inline src/dnscachedresolver.cpp]: You can `reserve` this to `dns_servers.size()` so you might as well.

File: src/dnscachedresolver.cpp [hunk 3]
```diff
@@ -767,8 +758,9 @@ DnsCachedResolver::DnsChannel* DnsCachedResolver::get_dns_channel()
   // Get the channel from the thread-local data, or create a new one if none
   // found.
   DnsChannel* channel = (DnsChannel*)pthread_getspecific(_thread_local);
+  size_t server_count = std::min((size_t)3u, _dns_servers.size());
```
Comment (bossmc) [inline src/dnscachedresolver.cpp]: Worth a log if we're dropping server entries?

File: src/dnscachedresolver.cpp [hunk 4]
```diff
@@ -791,6 +784,35 @@ DnsCachedResolver::DnsChannel* DnsCachedResolver::get_dns_channel()
...
+      {
+        memcpy(&ares_addr->addr.addr4, &server.addr.ipv4, sizeof(ares_addr->addr.addr4));
+      }
+      else
+      {
+        memcpy(&ares_addr->addr.addr6, &server.addr.ipv6, sizeof(ares_addr->addr.addr6));
+      }
+    }
+    
+    ares_set_servers(channel->channel, &(_ares_addrs[0]));
```
Comment (bossmc) [inline src/dnscachedresolver.cpp]: Why `&_ares_addrs[0]` that's the same as `_ares_addrs` (or probably `&_ares_addrs`... C++ be crazy, yo).

**Comment by bossmc** [comment]
Reviewed.

File: src/dnscachedresolver.cpp (hunk 4 above)
Comment (rkd-msw) [inline src/dnscachedresolver.cpp]: As discussed:

- it's a std::vector
- this is valid - pointers are only invalidated on reallocation (http://stackoverflow.com/a/8737812)

File: src/dnscachedresolver.cpp (hunk 3 above)
Comment (rkd-msw) [inline src/dnscachedresolver.cpp]: Sure.

File: src/dnscachedresolver.cpp (hunk 1 above)
Comment (rkd-msw) [inline src/dnscachedresolver.cpp]: Nope - and this preserves existing behaviour. (I tested with two invalid addresses when I wrote the "three DNS servers" code.)

File: src/dnscachedresolver.cpp (hunk 4 above)
Comment (bossmc) [inline src/dnscachedresolver.cpp]: Yep you're right, I was assuming that `_ares_addrs` was an array.  You can use [data()](http://www.cplusplus.com/reference/vector/vector/data/) rather that depending on the dereference/reference dance.

**Comment by rkd-msw** [comment]
OK, I've made your suggested markups, and also decided to have three constructors (IP46Address list, std::string list, std::string) rather than factory methods, just to reduce the amount of updating I have to do.

**Pull request merged**